src/Interpreters/Cache/WriteBufferToFileSegment.cpp

```diff
@@ -110,14 +110,11 @@ void WriteBufferToFileSegment::nextImpl()
 
 std::unique_ptr<ReadBuffer> WriteBufferToFileSegment::getReadBufferImpl()
 {
+    /** Finalize here and we don't need to finalize in the destructor,
+      * because in case destructor called without `getReadBufferImpl` called, data won't be read.
+      */
     finalize();
     return std::make_unique<ReadBufferFromFile>(file_segment->getPath());
 }
 
-WriteBufferToFileSegment::~WriteBufferToFileSegment()
-{
-    /// To be sure that file exists before destructor of segment_holder is called
-    WriteBufferFromFileDecorator::finalize();
```

Comment:
It seems after another changes we can remove finalization in the destructor.

We have to be sure that file exists in `downloaded_size > 0`, and we call `file_segment->setDownloadedSize(bytes_to_write);` after `impl->next()` in `WriteBufferToFileSegment::nextImpl()`, so in destructor eather any data is written or `downloaded_size == 0`

https://github.com/ClickHouse/ClickHouse/blob/6d544d94174fa1b0b320323a6942108f3887a98c/src/Interpreters/Cache/Metadata.cpp#L947-L953

https://github.com/ClickHouse/ClickHouse/blob/6d544d94174fa1b0b320323a6942108f3887a98c/src/Interpreters/Cache/FileSegment.cpp#L828-L840

Reply:
PR https://github.com/ClickHouse/ClickHouse/pull/63348 changes the code with `file_segment->setDownloadedSize`, but logic remains the same